neothesia/src/scene/playing_scene/midi_player.rs

```diff
@@ -198,69 +252,202 @@ pub enum MidiEventSource {
 struct UserPress {
     timestamp: Instant,
     note_id: u8,
+    time_key_up: Instant,
+    occurrence: usize,
 }
 
-#[derive(Debug)]
 pub struct PlayAlong {
     user_keyboard_range: piano_math::KeyboardRange,
 
-    required_notes: HashSet<u8>,
-
+    required_notes: VecDeque<UserPress>,
```

Comment:
It seems like you are never using the ring buffer aspect of the VecDeque, let's just use a regular `Vec` and `push()` instead of `push_back()`
```suggestion
    required_notes: Vec<UserPress>,
```

Reply:
Oh, actually it looks like like this should also be a `HashMap<NoteId, UserPress>`, right?

same way as https://github.com/PolyMeilex/Neothesia/pull/172/files#r1614080982

Reply:
I just kept the .push_back() i am not sure a double-ended is needed anymore in the code now, it is using only retain, it doesn't handle indexes and positions anymore.. will test with single vecs and if gameplay turns okay i will push